google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/CloudSdkAppEngineHelper.java

```diff
@@ -79,61 +139,148 @@ public File defaultDockerfile(DeploymentArtifactType deploymentArtifactType) {
 
   @NotNull
   @Override
-  public AppEngineDeployAction createDeploymentAction(
+  public CancellableRunnable createDeployRunner(
       LoggingHandler loggingHandler,
-      Project project,
       File artifactToDeploy,
       AppEngineDeploymentConfiguration deploymentConfiguration,
-      DeploymentOperationCallback deploymentCallback) throws IllegalArgumentException {
-    DeploymentArtifactType artifactType = DeploymentArtifactType.typeForPath(artifactToDeploy);
-    if (artifactType == DeploymentArtifactType.UNKNOWN) {
-      throw new IllegalArgumentException(artifactToDeploy.getPath() + " is not a support artifact "
-          + "type for automatic deployment");
-    }
-    return new AppEngineDeployAction(
+      DeploymentOperationCallback callback) {
+    AppEngineDeploy deploy = new AppEngineDeploy(
         this,
         loggingHandler,
-        project,
-        artifactToDeploy,
         deploymentConfiguration,
-        wrapCallbackForUsageTracking(deploymentCallback, ConfigType.AUTO, artifactType)
-    );
+        wrapCallbackForUsageTracking(callback, deploymentConfiguration, artifactToDeploy));
+
+    if (environment == Environment.APP_ENGINE_STANDARD) {
+      AppEngineStandardStage standardStage = new AppEngineStandardStage(
+          this,
+          loggingHandler,
+          artifactToDeploy);
+
+      return new AppEngineStandardDeployRunner(deploy, standardStage);
+    } else {
+      AppEngineFlexibleStage flexibleStage = new AppEngineFlexibleStage(
+          this,
+          loggingHandler,
+          artifactToDeploy,
+          deploymentConfiguration);
+
+      return new AppEngineFlexibleDeployRunner(deploy, flexibleStage);
+    }
   }
 
-  @NotNull
   @Override
-  public AppEngineStopAction createStopAction(
+  public File createStagingDirectory(LoggingHandler loggingHandler) throws IOException {
+    File stagingDirectory = FileUtil.createTempDirectory(
+        "gae-mvm" /* prefix */,
```

Comment:
maybe we should change this prefix to something like gae-flex-staging-project-service-version